### PR TITLE
JDK-8318485: Narrow klass shift should be zero if encoding range extends to 0x1_0000_0000

### PR DESCRIPTION
**Pull request by tstuefe**
See JBS issue.

The real fix, trivial, is the comparison change in compressedKlass.cpp.

This patch then also adds a test that checks that for a given (forced) class space location the VM chose the correct encoding scheme. In this case, if class space nestles below and ends at 4G, we should use a shift of 0.

The test can be easily extended to test more narrow Klass encoding schemes. Note that the test is skipped if the VM could not map at the expected address to begin with.

Finally, to analyze test errors, I print out the mappings at the force address if metaspace logging is debug.

Tests: manual, GHA. x86 GHA errors unrelated.

/label remove core-libs
/label remove hotspot
/label add hotspot-runtime

Attention @iklam and @calvinccheung

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8318485](https://bugs.openjdk.org/browse/JDK-8318485): Narrow klass shift should be zero if encoding range extends to 0x1_0000_0000 (**Bug** - P4)

### Reviewers
 * [Calvin Cheung](https://openjdk.org/census#ccheung) (@calvinccheung - **Reviewer**)
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16261/head:pull/16261` \
`$ git checkout pull/16261`

Update a local copy of the PR: \
`$ git checkout pull/16261` \
`$ git pull https://git.openjdk.org/jdk.git pull/16261/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16261`

View PR using the GUI difftool: \
`$ git pr show -t 16261`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16261.diff">https://git.openjdk.org/jdk/pull/16261.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16261#issuecomment-1772121100)

### CODE REVIEW

File: src/hotspot/share/memory/metaspace.cpp [hunk 1]
```diff
@@ -785,6 +785,11 @@ void Metaspace::global_initialize() {
       if (rs.is_reserved()) {
         log_info(metaspace)("Successfully forced class space address to " PTR_FORMAT, p2i(base));
       } else {
+        LogTarget(Debug, metaspace) lt;
+        if (lt.is_enabled()) {
+          LogStream ls(lt);
+          os::print_memory_mappings((char*)base, size, &ls);
```
Comment (calvinccheung) [inline src/hotspot/share/memory/metaspace.cpp:791]: I saw the above log message as follows:
```
java -XX:+UnlockDiagnosticVMOptions -XX:CompressedClassSpaceBaseAddress=0xa0000000 -XX:CompressedClassSpaceSize=1g -Xlog:metaspace=debug -Xlog:os+map=trace -Xshare:off -version
[0.001s][info][metaspace]  - commit_granule_bytes: 65536.
[0.001s][info][metaspace]  - commit_granule_words: 8192.
[0.001s][info][metaspace]  - virtual_space_node_default_size: 8388608.
[0.001s][info][metaspace]  - enlarge_chunks_in_place: 1.
[0.001s][info][metaspace]  - use_allocation_guard: 0.
[0.036s][debug][metaspace] Range [a0000000-e0000000) contains: 
[0.036s][debug][metaspace] a0000000-120000000 rw-p 00000000 00:00 0 
[0.036s][debug][metaspace] 
Error occurred during initialization of VM
CompressedClassSpaceBaseAddress=0x00000000a0000000 given, but reserving class space failed.
```
Is it easy to get rid of the blank line before the line begins with "Error occurred..."?

File: test/hotspot/jtreg/runtime/CompressedOops/CompressedClassPointersEncodingScheme.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+    final static long K = 1024;
+    final static long M = K * 1024;
+    final static long G = M * 1024;
+    public static void main(String[] args) throws Exception {
+        // Test ccs nestling right at the end of the 4G range
+        // Expecting base=0, shift=0
+        test(4 * G - 128 * M, 128 * M, 0, 0);
+
+        // add more...
```
Comment (calvinccheung) [inline test/hotspot/jtreg/runtime/CompressedOops/CompressedClassPointersEncodingScheme.java:78]: The following works on my linux dev host.
```
        test(0xa0000000L, 1 * G, 0, 0);
        test(0x100000000L, 2 * G, 0, 3);
```
Maybe you can consider adding them to the test?

**Review by calvinccheung — COMMENTED** [review]
Thanks for fixing this. I've tried your patch on the CompressedClassPointers.java test. It also fixes the issue in [JDK-8317610](https://bugs.openjdk.org/browse/JDK-8317610). Just have a couple of minor comments.

File: src/hotspot/share/memory/metaspace.cpp (hunk 1 above)
Comment (tstuefe) [reply src/hotspot/share/memory/metaspace.cpp:791]: Okay.

File: test/hotspot/jtreg/runtime/CompressedOops/CompressedClassPointersEncodingScheme.java (hunk 2 above)
Comment (tstuefe) [reply test/hotspot/jtreg/runtime/CompressedOops/CompressedClassPointersEncodingScheme.java:78]: I'd rather expand this in a separate RFE. Second one *should* work on all platforms, but I am not 100% sure there are some that would encode this with base=4g shift=0 instead. Probably not, but its better to stay focussed on one issue, its also better for backporting.

**Comment by tstuefe** [comment]
Thanks @calvinccheung . I fixed the output as requested.

**Review by calvinccheung — APPROVED** [review]
Thanks for the update. It looks good.

**Review by iklam — APPROVED** [review]
LGTM

**Comment by tstuefe** [comment]
Thanks @calvinccheung and @iklam .

/integrate

**Pull request closed without merging**